Applications/Spire/Include/Spire/Ui/NumericFilterPanel.hpp

```diff
@@ -0,0 +1,274 @@
+
+      /** Returns the default maximum value. */
+      boost::optional<Type> get_default_max_value() const;
+
+      /** Sets the default minimum and maximum value. */
+      void set_default_value(const boost::optional<Type>& default_min,
+        const boost::optional<Type>& default_max);
+
+      /** Connects a slot to the filter signal. */
+      boost::signals2::connection connect_filter_signal(
+        typename const FilterSignal::slot_type& slot) const;
+
+    protected:
+      bool event(QEvent* event) override;
+
+    private:
+      mutable FilterSignal m_filter_signal;
+      std::shared_ptr<OptionalNumericModel> m_min_model;
+      std::shared_ptr<OptionalNumericModel> m_max_model;
+      boost::optional<Type> m_default_min_value;
+      boost::optional<Type> m_default_max_value;
+      boost::optional<Type> m_min_value;
+      boost::optional<Type> m_max_value;
+      FilterPanel* m_filter_panel;
+
+      NumericBox* make_input_field(std::shared_ptr<OptionalNumericModel> model);
+      QHBoxLayout* make_row_layout(QString label_name, NumericBox* input_field);
+      void reset();
+  };
+
+  template<typename T>
+  NumericFilterPanel<T>::NumericFilterPanel(
+    boost::optional<Type> default_min_value,
+    boost::optional<Type> default_max_value,
+    QString tilte, QWidget* parent)
+    : NumericFilterPanel(std::make_shared<LocalOptionalNumericModel>(),
+      std::make_shared<LocalOptionalNumericModel>(), default_min_value,
+      default_max_value, title, parent) {}
+
+  template<typename T>
+  NumericFilterPanel<T>::NumericFilterPanel(
+      std::shared_ptr<OptionalNumericModel> min_model,
+      std::shared_ptr<OptionalNumericModel> max_model,
+      boost::optional<Type> default_min_value,
+      boost::optional<Type> default_max_value,
+      QString title, QWidget* parent)
+      : QWidget(parent),
+        m_min_model(std::move(min_model)),
+        m_max_model(std::move(max_model)),
+        m_default_min_value(std::move(default_min_value)),
+        m_default_max_value(std::move(default_max_value)) {
+    if(m_default_min_value && m_default_max_value &&
+        *m_default_min_value > *m_default_max_value) {
+      std::swap(m_default_min_value, m_default_max_value);
+    }
+    auto layout = new QVBoxLayout(this);
+    layout->setSpacing(0);
+    layout->setContentsMargins({});
+    auto min_input = make_input_field(m_min_model);
+    auto min_layout = make_row_layout(tr("Min"), min_input);
+    layout->addLayout(min_layout);
+    layout->addSpacing(scale_height(10));
+    auto max_input = make_input_field(m_max_model);
+    auto max_layout = make_row_layout(tr("Max"), max_input);
+    layout->addLayout(max_layout);
+    m_filter_panel = new FilterPanel(std::move(title), this, parent);
+    m_filter_panel->connect_reset_signal([=] {
+      reset();
+    });
+    min_input->connect_submit_signal([=] (const auto& value) {
+      if(m_max_value && value && *value > *m_max_value) {
```

Comment:
The constructor shall only be used to initialize an object, do not implement anything inside of it other than what is needed to fully initialize an object. It is very hard to reason about code that mixes in several different responsibilities and layers together.